### PR TITLE
Support Ruby 2.4.

### PR DESCRIPTION
**Pull request by tgxworld**
`ruby script/bench.rb -u -n -b 3`

# 2.4.1
```
---
categories_admin:
  50: 12
  75: 12
  90: 14
  99: 22
home_admin:
  50: 16
  75: 16
  90: 18
  99: 23
topic_admin:
  50: 12
  75: 13
  90: 14
  99: 19
categories:
  50: 17
  75: 18
  90: 21
  99: 27
home:
  50: 13
  75: 13
  90: 15
  99: 20
topic:
  50: 10
  75: 11
  90: 11
  99: 16
timings:
  load_rails: 2365
ruby-version: 2.4.1-p111
rss_kb: 201288
pss_kb: 139086
virtual: physical
architecture: amd64
operatingsystem: Ubuntu
memorysize: 15.59 GB
kernelversion: 4.4.0
physicalprocessorcount: 1
processor0: Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz
rss_kb_31987: 304992
pss_kb_31987: 246176
rss_kb_32051: 304356
pss_kb_32051: 245368
rss_kb_32174: 303300
pss_kb_32174: 239080
```

# 2.3.3
```
---
categories_admin:
  50: 13
  75: 14
  90: 17
  99: 28
home_admin:
  50: 16
  75: 17
  90: 18
  99: 26
topic_admin:
  50: 12
  75: 13
  90: 14
  99: 23
categories:
  50: 18
  75: 19
  90: 19
  99: 29
home:
  50: 14
  75: 15
  90: 16
  99: 25
topic:
  50: 11
  75: 12
  90: 12
  99: 20
timings:
  load_rails: 2563
ruby-version: 2.3.3-p222
rss_kb: 234160
pss_kb: 165813
virtual: physical
architecture: amd64
operatingsystem: Ubuntu
memorysize: 15.59 GB
kernelversion: 4.4.0
physicalprocessorcount: 1
processor0: Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz
rss_kb_26070: 368484
pss_kb_26070: 298182
rss_kb_26143: 372212
pss_kb_26143: 301783
rss_kb_26255: 367932
pss_kb_26255: 297356
```

### CODE REVIEW

**Comment by tgxworld** [comment]
One of the plugin test is still using `FakeWeb` which explains the test failure on travis.

@SamSaffron  Can you review thank you!

File: Gemfile.lock [hunk 1]
```diff
@@ -37,6 +37,8 @@ GEM
       minitest (~> 5.1)
       thread_safe (~> 0.3, >= 0.3.4)
       tzinfo (~> 1.1)
+    addressable (2.5.1)
```
Comment (tgxworld) [inline Gemfile.lock]: @SamSaffron  I know you hate addressable for its memory footprint but we'll only be installing this in the test environment. `FakeWeb` hasn't been updated for a couple of years and doesn't work with Ruby 2.4.1 at all. I've had to switch to `WebMock` which introduces this dependency.

File: app/models/user_warning.rb [hunk 2]
```diff
@@ -1,4 +1,4 @@
-class Warning < ActiveRecord::Base
+class UserWarning < ActiveRecord::Base
```
Comment (tgxworld) [inline app/models/user_warning.rb]: We can't use `Warning` anymore

https://ruby-doc.org/core-2.4.0/Warning.html

File: Gemfile.lock [hunk 3]
```diff
@@ -69,6 +71,8 @@ GEM
     coderay (1.1.1)
     concurrent-ruby (1.0.5)
     connection_pool (2.2.0)
+    crack (0.4.3)
```
Comment (SamSaffron) [inline Gemfile.lock]: YAY CRACK

**Comment by SamSaffron** [comment]
specs are failing, but once that is sorted @tgxworld  I am delighted with this change. We should get 2.4 on meta ASAP and graph memory changes.

**Pull request merged**